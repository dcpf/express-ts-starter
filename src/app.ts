import express, { NextFunction, Request, Response } from 'express';
import path from 'path';
import cookieParser from 'cookie-parser';
import cors from 'cors';
import helmet from 'helmet';
import logger from 'morgan';

import apiRouter from './routes/apiRouter';
import viewRouter from './routes/viewRouter';
import { ApiError } from './error/ApiError';

const BASE_API_PATH = '/api';

export const app = express();

app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'ejs');

app.use(helmet());
app.use(cors());
app.use(logger('combined'));
app.use(express.json());
app.use(express.urlencoded({ extended: false }));
app.use(cookieParser());
app.use(express.static(path.join(__dirname, 'public')));

app.use(BASE_API_PATH, apiRouter);
app.use('/', viewRouter);

// api error handler
app.use((err: Error, req: Request, res: Response, next: NextFunction) => {
  if (err instanceof ApiError) {
    console.error(err.statusCode, err.message);
    res.status(err.statusCode).json({ message: err.message });
    return;
  }
  if (req.url.startsWith(`${BASE_API_PATH}/`)) {
    console.error(err);
    res.status(500).json({ message: err.message });
    return;
  }
  next(err);
});

// view error handler
app.use((err: Error, _req: Request, res: Response, _next: NextFunction) => {
  // set locals
  res.locals.message = err.message;
  res.locals.error = err;

  // render the error page
  res.status(500);
  res.render('error');
});
