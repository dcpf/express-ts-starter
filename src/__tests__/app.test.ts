import request from 'supertest';
import { app } from '../app';

describe('App tests', () => {
  describe('View tests', () => {
    test('home page view', async () => {
      const response = await request(app).get('/');
      expect(response.statusCode).toBe(200);
      expect(response.text).toMatchSnapshot();
    });
  });

  describe('API tests', () => {
    it('get all users', async () => {
      const response = await request(app).get('/api/users');
      expect(response.statusCode).toBe(200);
      expect(response.body).toEqual([
        { email: 'bob@example.com', id: '1', name: 'Bob Example' },
        { email: 'd.duck@wb.com', id: '2', name: 'Daffy Duck' },
        { email: 'dirk@rush.com', id: '3', name: 'Geddy Lee' },
        { email: 'p.mccartney@beatles.com', id: '4', name: 'Paul McCartney' },
      ]);
    });

    it('get user', async () => {
      const response = await request(app).get('/api/users/3');
      expect(response.statusCode).toBe(200);
      expect(response.body).toEqual({
        email: 'dirk@rush.com',
        id: '3',
        name: 'Geddy Lee',
      });
    });

    it('get user - not found', async () => {
      const response = await request(app).get('/api/users/5');
      expect(response.statusCode).toBe(404);
      expect(response.body).toEqual({ message: 'No user found for id 5' });
    });
  });
});
