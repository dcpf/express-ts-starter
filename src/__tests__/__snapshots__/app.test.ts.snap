// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests View tests home page view 1`] = `
"<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta content='width=device-width' name='viewport' />
  <title>Express with TypeScript Starter</title>
  <link rel="stylesheet" type="text/css" href="/stylesheets/styles.css" />
</head>

<body>
  Hello World!
</body>

</html>"
`;
